### ISSUE TITLE
Some edge cases in `email.utils.parsedate_to_datetime` seem to differ from RFC2822 spec

### ISSUE THREAD

**Issue by ariebovenberg** [python/cpython#126845]
# Bug report

### Bug description:

While tinkering around with `email.utils.parsedate_to_datetime`, I found some behavior that may be worth adjusting.

##  1. low-number years aren't handled according to spec:

> The year is any numeric year 1900 or later. [section 3.3]
>
> [section 4.3] The syntax for the obsolete date format allows a 2 digit year.
> [..]
> Where a two or three digit year occurs in a date, the year is to be
> interpreted as follows: If a two digit year is encountered whose
> value is between 00 and 49, the year is interpreted by adding 2000,
> ending up with a value between 2000 and 2049.  If a two digit year is
> encountered with a value between 50 and 99, or any three digit year
> is encountered, the year is interpreted by adding 1900.

```python
>>> parsedate_to_datetime("Sat, 15 Aug 0001 23:12:09 +0500")
datetime.datetime(2001, 8, 15, 23, 12, 9, ...)
```

expected: either year 1, or a parsing failure. Neither the new or old format interpret 4-digit years this way.

## 2. offset minutes larger than 59 don't lead to parsing failure

```python
>>> parsedate_to_datetime('Sat, 15 Aug 0001 23:12:09 +0590')
datetime.datetime(2001, 8, 15, 23, 12, 9, tzinfo=datetime.timezone(datetime.timedelta(seconds=23400)))
```

expected: parse failure. Instead, the "90 minutes" component is parsed without issue (0590 being equal to 0630). The spec is actually not explicit about this, although _"A date-time specification MUST be semantically valid"_. Note that a "90" value as minute in the time component _does_ give the appropriate parsing failure. 

Note: `datetime.fromisoformat()` has the same behavior. Also in this case, I can't determine whether ISO8601 explicitly disallows it. RFC3339 _is_ clear on _disallowing_ this.

## 3. Invalid day-of-week doesn't lead to parsing failure

```python
>>> parsedate_to_datetime('Sun, 15 Aug 0001 23:12:09 +0520')  # actually a saturday
```

expected: parsing failure

> A date-time specification MUST be semantically valid.  That is, the
> day-of-the-week (if included) MUST be the day implied by the date,

## 4. Non-ASCII digits don't lead to parsing failure

If I'm reading the RFC correctly, only ASCII characters are valid.

```python
>>> parsedate_to_datetime('Sat, 15 Aug 01 𝟚𝟛:𝟝𝟛:𝟛𝟛 +0500')  # note the fancy numbers
datetime.datetime(2001, 8, 15, 23, 53, 33, ...)
```

expected: parsing failure

## 5. Handling of the `-0000` case may be inconsistent with drive to eliminate the practice of "naive UTC" datetimes.

Lately, the `datetime` module appears to discourage the usage of naive datetimes to mean UTC, as evidenced by the deprecation of `utcnow()` and other methods.

However, `parsedate_to_datetime` will return a _naive_ datetime in the `-0000` case.

```python
>>> parsedate_to_datetime("Sat, 15 Aug 01 23:53:33 -0000")
datetime.datetime(2001, 8, 15, 23, 53, 33)
```

expected: `tzinfo=UTC`

The spec says:

> "-0000" also indicates Universal Time, it is
> used to indicate that the time was generated on a system that may be
> in a local time zone other than Universal Time and therefore
> indicates that the date-time contains no information about the local
> time zone.

The spec again is a bit fuzzy, but my reading here is that `-0000` means "UTC, with no offset known". In contrast, `+0000`  means "UTC offset known to be 0". My impression would be that only _omission_ of the offset should result in a naive datetime. What do you think?

### CPython versions tested on:

3.13

### Operating systems tested on:

macOS

edit: typo

### Linked PRs
* gh-134311
* gh-134350
* gh-134438

**Comment by encukou** [comment]
Thanks for the report!

[`parsedate_to_datetime` docs](https://docs.python.org/3/library/email.utils.html#email.utils.parsedate_to_datetime) refer to `parsedate()`, whose docs say:
> some mailers don’t follow that format as specified, so `parsedate()` tries to guess correctly in such cases.

That is, these functions parse but *do not validate* their input. Full validation is left for a third-party library. (Python *could* add it, taking a *standard* as an argument and validating based on that, but I don't think anyone's interested in maintaining it in stdlib.)

With that, IMO:
1. is a bug, that should be year 1. Do you want to send a PR?
2. is fine
3. is fine
4. is fine
5. is fine -- the docs cover this case explicitly. (AFAICS, RFC 2822 does not allow omitting the zone, so `-0000` is the way to indicate a naive datetime.)

**Comment by ariebovenberg** [comment]
@encukou thanks—sorry I missed that context. I'm happy to submit a PR for the year issue.

**Comment by bitdancer** [comment]
For 1, yes, there is a bug here.  A three digit year should have 1900 added, which currently doesn't happen.  Will fixing that break anyone's code?  I'm guessing not, so I'm OK with fixing it.  But we should probably not backport it.

As for the four digit year, yes that should be an error: only dates 1900 and later are technically valid.  Changing that probably *would* break someones code, so I'm OK with returning four digit years before 1900 as written, since it is a reasonable Postel recovery (and actually makes more sense than restricting it to post 1900, IMO).  That will change the current behavior for dates in the range 0000-0099, but again that seems unlikely to break anyone's code.

For 2 and 3 we'd ideally raise an error, but that might break existing code.

4 depends on whether `utf8` is true or not.  Accepting non-ascii unicode numbers is a reasonable Postel recovery, so I don't think we should change that.  The new email API should (I think, I haven't tested it) register a defect for that case.

5 follows the RFC, as noted by encukou.

I would like to see the new API register defects for 1, 2, and 3, but doing that would require a fair bit of code tweaking, so we probably shouldn't waste time on it, at least currently.

**Comment by bitdancer** [comment]
On the (reworked) PR #150864, Serhiy pointed out that the RFC actually says "49" for the cutoff, not the "68" the code currently uses.  After a little research I see what happened there.  The original fix (not by me, though I committed it) used the *posix* standard definition for two digit years.  Some more (AI assisted) research indicates that there is considerable variation on how other email parsers handle this.    Perl follows posix for the most part.  More modern software (including later perl libraries) allows one to set the pivot.  For email an ability to set the pivot should not be particularly important since four digit dates are now required. Thunderbird defers to the OS (glibc) locale settings, where Windows uses 49 and Linux uses 68. Outlook of course uses 49.  Apple mail, equally of course, rolls its own, and uses 29 as the cutoff (30-99 is 1930-1999).  Gmail uses a sliding window depending on the current date.

In short, there is no good answer here, so I'm inclined to leave this as it is and just update the comment with an explanation.  There is certainly an argument for following the standard, and if this hadn't been our reality since 2010 I'd do that.  Maybe we still should?  It's pretty much impossible to guess if this would break anyone's code.
